### PR TITLE
Moving funds timeout

### PR DESCRIPTION
**Pull request by lukasz-zimnoch**
Closes: #179 
Depends on: #208

### CODE REVIEW

File: solidity/contracts/bridge/Bridge.sol [hunk 1]
```diff
@@ -530,6 +533,16 @@ contract Bridge is Ownable, EcdsaWalletOwner {
         );
     }
 
+    /// @notice Notifies about a timed out moving funds process. Terminates
+    ///         the wallet in result.
```
Comment (pdyraga) [inline solidity/contracts/bridge/Bridge.sol:537]:
```suggestion
    /// @notice Notifies about a timed out moving funds process. Terminates
    ///         the wallet and slashes signing group members as a result.
```

File: solidity/contracts/bridge/Bridge.sol [hunk 2]
```diff
@@ -190,6 +192,7 @@ contract Bridge is Ownable, EcdsaWalletOwner {
         self.redemptionTxMaxFee = 10000; // 10000 satoshi
         self.redemptionTimeout = 172800; // 48 hours
         self.movingFundsTxMaxTotalFee = 10000; // 10000 satoshi
+        self.movingFundsTimeout = 7 days;
```
Comment (pdyraga) [inline solidity/contracts/bridge/Bridge.sol:195]: 👌🏻 Yeah, let's not be overambitious at the beginning.

File: solidity/contracts/bridge/Bridge.sol [hunk 3]
```diff
@@ -963,13 +976,17 @@ contract Bridge is Ownable, EcdsaWalletOwner {
     ///         transaction fee that is acceptable in a single moving funds
     ///         transaction. This is a _total_ max fee for the entire moving
     ///         funds transaction.
+    /// @return movingFundsTimeout Time after which the moving funds process
```
Comment (pdyraga) [inline solidity/contracts/bridge/Bridge.sol:979]: It is in seconds, right? Worth mentioning it in the docs.

File: solidity/contracts/bridge/MovingFunds.sol [hunk 4]
```diff
@@ -349,4 +351,19 @@ library MovingFunds {
 
         return (targetWalletsHash, outputsTotalValue);
     }
+
+    /// @notice Notifies about a timed out moving funds process. Terminates
+    ///         the wallet in result.
```
Comment (pdyraga) [inline solidity/contracts/bridge/MovingFunds.sol:356]:
```suggestion
    /// @notice Notifies about a timed out moving funds process. Terminates
    ///         the wallet in the result.
```

File: solidity/contracts/bridge/Wallets.sol [hunk 5]
```diff
@@ -624,4 +624,34 @@ library Wallets {
...
+        require(
+            wallet.state == WalletState.MovingFunds,
+            "ECDSA wallet must be in MovingFunds state"
+        );
+
+        require(
+            /* solhint-disable-next-line not-rely-on-time */
+            block.timestamp >
+                wallet.movingFundsRequestedAt + self.movingFundsTimeout,
+            "Moving funds has not timed out"
```
Comment (pdyraga) [inline solidity/contracts/bridge/Wallets.sol:649]:
```suggestion
            "Moving funds has not timed out yet"
```

File: solidity/contracts/bridge/Wallets.sol [hunk 6]
```diff
@@ -624,4 +624,34 @@ library Wallets {
...
+        );
```
Comment (pdyraga) [inline solidity/contracts/bridge/Wallets.sol:650]: Based on [this conversation](https://github.com/keep-network/tbtc-v2/pull/208/files#discussion_r854996501), I think this part should be in `MovingFunds` library. I think we also don't need `notifyWalletMovingFundsTimeout` in `Wallets` at all. It is fine if we call `terminateWallet` directly. IMO, slashing in `MovingFunds` is fine as well.

File: solidity/test/bridge/Bridge.MovingFunds.test.ts [hunk 7]
```diff
@@ -1455,6 +1455,149 @@ describe("Bridge - Moving funds", () => {
...
+            .to.emit(bridge, "MovingFundsTimedOut")
+            .withArgs(ecdsaWalletTestData.pubKeyHash160)
+        })
+
+        it("should slash wallet operators", async () => {
+          // TODO: Implementation once slashing is integrated.
+        })
+
+        it("should reward the notifier", async () => {
+          // TODO: Implementation once slashing is integrated.
```
Comment (pdyraga) [inline solidity/test/bridge/Bridge.MovingFunds.test.ts:1540]: We will need to add governable slashing amount and notifier multiplier parameters to the contract.

Comment (pdyraga) [reply solidity/test/bridge/Bridge.MovingFunds.test.ts:1540]: Can happen in the PR where we'll integrate with https://github.com/keep-network/keep-core/pull/2942, not necessarily this one - leaving it up to you.

File: solidity/contracts/bridge/Wallets.sol [hunk 8]
```diff
@@ -624,4 +624,34 @@ library Wallets {
...
+
+        terminateWallet(self, walletPubKeyHash);
+
+        // TODO: Perform slashing of wallet operators, reward the notifier
+        //       using seized amount, and add unit tests for that.
```
Comment (pdyraga) [inline solidity/contracts/bridge/Wallets.sol:655]: If you review https://github.com/keep-network/keep-core/pull/2942 we can do it now 😉 

It would be great to test it with a real `WalletRegistry` and `TokenStaking`. We can have a separate test file and setup for that to do not complicate tests covering corner cases but it's definitely worth having the integration unit-tested as well. We will at least see the real gas cost.

File: solidity/test/bridge/Bridge.MovingFunds.test.ts [hunk 9]
```diff
@@ -1455,6 +1455,149 @@ describe("Bridge - Moving funds", () => {
...
+        })
+      })
+
+      context("when the moving funds process has not timed out", () => {
+        it("should revert", async () => {
+          await expect(
+            bridge.notifyMovingFundsTimeout(ecdsaWalletTestData.pubKeyHash160)
+          ).to.be.revertedWith("Moving funds has not timed out")
+        })
+      })
```
Comment (tomaszslabon) [inline solidity/test/bridge/Bridge.MovingFunds.test.ts:1550]: In most other test cases where we test timeout notifications, we increase time by the `timeout` value minus one second.
We could consider using this pattern here.

File: solidity/test/bridge/Bridge.MovingFunds.test.ts (hunk 7 above)
Comment (lukasz-zimnoch) [reply solidity/test/bridge/Bridge.MovingFunds.test.ts:1592]: Yeah sure but I prefer to handle all slashing separately. We have TODOs that capture all places.

File: solidity/contracts/bridge/Wallets.sol (hunk 8 above)
Comment (lukasz-zimnoch) [reply solidity/contracts/bridge/Wallets.sol]: I can review it but I prefer to implement all the slashing/seizing separately. Agreed about the unit tests, we can add them once the integration is ready in the code.

File: solidity/contracts/bridge/Bridge.sol (hunk 1 above)
Comment (lukasz-zimnoch) [reply solidity/contracts/bridge/Bridge.sol]: See https://github.com/keep-network/tbtc-v2/pull/212/commits/4af794e278898ada7a640d7e6124ef5cef16696c

File: solidity/contracts/bridge/Bridge.sol (hunk 3 above)
Comment (lukasz-zimnoch) [reply solidity/contracts/bridge/Bridge.sol:984]: See https://github.com/keep-network/tbtc-v2/pull/212/commits/4af794e278898ada7a640d7e6124ef5cef16696c

File: solidity/contracts/bridge/MovingFunds.sol (hunk 4 above)
Comment (lukasz-zimnoch) [reply solidity/contracts/bridge/MovingFunds.sol]: See https://github.com/keep-network/tbtc-v2/pull/212/commits/4af794e278898ada7a640d7e6124ef5cef16696c

File: solidity/contracts/bridge/Wallets.sol (hunk 5 above)
Comment (lukasz-zimnoch) [reply solidity/contracts/bridge/Wallets.sol]: See https://github.com/keep-network/tbtc-v2/pull/212/commits/4af794e278898ada7a640d7e6124ef5cef16696c

File: solidity/test/bridge/Bridge.MovingFunds.test.ts (hunk 9 above)
Comment (lukasz-zimnoch) [reply solidity/test/bridge/Bridge.MovingFunds.test.ts]: Yeah, see https://github.com/keep-network/tbtc-v2/pull/212/commits/72832423a72d567caf23f79dc53216c6f6245a8b

File: solidity/contracts/bridge/MovingFunds.sol [hunk 10]
```diff
@@ -426,4 +428,38 @@ library MovingFunds {
...
+        BridgeState.Storage storage self,
+        bytes20 walletPubKeyHash
+    ) external {
+        Wallets.Wallet storage wallet = self.registeredWallets[
+            walletPubKeyHash
+        ];
+
+        require(
+            wallet.state == Wallets.WalletState.MovingFunds,
+            "ECDSA wallet must be in MovingFunds state"
```
Comment (pdyraga) [inline solidity/contracts/bridge/MovingFunds.sol:448]:
```suggestion
            "Wallet must be in MovingFunds state"
```

Non-blocking this PR but would be great to align revert messages about wallet states:
- In `Deposit.revealDeposit` we say `Wallet is not in Live state`
- In `Redemption.requestRedemption` we say `Wallet must be in Live state`
- In `Redemption.submitRedemptionProof` we say `Wallet must be in Live or MovingFunds state`
- In `Redemption.notifyRedemptionTimeout` we say `The wallet must be in Live, MovingFunds or Terminated state` <-- something to align as well
- In `Sweep.resolveSweepingWallet` we say `Wallet must be in Live or MovingFunds state`
- In `Fraud.submitFraudChallenge` we say `Wallet is neither in Live nor MovingFunds state`

**Review by pdyraga — APPROVED** [review]

**Comment by pdyraga** [comment]
Slither @lukasz-zimnoch 😬
